response.go

```diff
@@ -0,0 +1,171 @@
+package httpapi
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"os"
+	"strings"
+
+	files "github.com/ipfs/go-ipfs-files"
+)
+
+type trailerReader struct {
+	resp *http.Response
+}
+
+func (r *trailerReader) Read(b []byte) (int, error) {
+	n, err := r.resp.Body.Read(b)
+	if err != nil {
+		if e := r.resp.Trailer.Get("X-Stream-Error"); e != "" {
+			err = errors.New(e)
+		}
+	}
+	return n, err
+}
+
+func (r *trailerReader) Close() error {
+	return r.resp.Body.Close()
+}
+
+type Response struct {
+	Output io.ReadCloser
+	Error  *Error
+
+	drainOutput bool
+}
+
+func (r *Response) Close() error {
+	if r.Output != nil {
+
+		// always drain output (response body)
+		var err1 error
+		if r.drainOutput {
+			_, err1 = io.Copy(ioutil.Discard, r.Output)
+		}
+		err2 := r.Output.Close()
+		if err1 != nil {
+			return err1
+		}
+		return err2
+	}
+	return nil
+}
+
+func (r *Response) Decode(dec interface{}) error {
+	defer r.Close()
+	if r.Error != nil {
+		return r.Error
+	}
+
+	n := 0
+	var err error
+	for {
+		err = json.NewDecoder(r.Output).Decode(dec)
+		if err != nil {
+			break
+		}
+		n++
+	}
+	if n > 0 && err == io.EOF {
+		err = nil
+	}
+	return err
+}
+
+type Error struct {
+	Command string
+	Message string
+	Code    int
+}
+
+func (e *Error) Error() string {
+	var out string
+	if e.Code != 0 {
+		out = fmt.Sprintf("%s%d: ", out, e.Code)
+	}
+	return out + e.Message
+}
+
+func (r *Request) Send(c *http.Client) (*Response, error) {
+	url := r.getURL()
+	req, err := http.NewRequest("POST", url, r.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	// Add any headers that were supplied via the RequestBuilder.
+	for k, v := range r.Headers {
+		req.Header.Add(k, v)
+	}
+
+	if fr, ok := r.Body.(*files.MultiFileReader); ok {
+		req.Header.Set("Content-Type", "multipart/form-data; boundary="+fr.Boundary())
+		req.Header.Set("Content-Disposition", "form-data: name=\"files\"")
+	}
+
+	resp, err := c.Do(req)
+	if err != nil {
+		return nil, err
+	}
+
+	contentType := resp.Header.Get("Content-Type")
+	parts := strings.Split(contentType, ";")
```

Comment:
It may be overkill, but it would be nice to use `mime.ParseMediaType` here (from the `mime` package).